### DISCUSSION TITLE
Should we switch away from Gitter

### DISCUSSION THREAD

**Discussion by StephanHoyer** [MithrilJS/mithril.js#2740] [General]
We discussed internally, if we should switch away from [gitter](https://gitter.im/mithriljs/mithril.js) as main chat platform. I think it would be better, to discuss this here.

What are possible alternatives and what are the pros and cons.

**Comment by panoply** [comment]
One of the main issues with moving away from Gitter and choosing an alternative like Discord is the character limits that are imposed. Linking a Flems in Discord is simply not possible.

**Reply by kevinfiol** [reply]
This is a great point. Discord is a popular choice among OSS projects these days, but if Flems links (or any link for that matter) breaks within Discord, that would be a major downside.

Also, I expressed this in chat but I believe Gitter actually has a lower barrier of entry than Discord for people to enter and see discussion. 

1. You don't need to be logged into Gitter to lurk the chatroom
2. Gitter requires a Github or Gitlab account, which most developers likely already have

**Reply by tbreuss** [reply]
Support for Flems is definitely a must-have.

**Comment by tbreuss** [comment]
Zulip was mentioned several times in the internal channel.

https://zulip.com/for/communities/

Pros:
- It is basically open source slack with subjects
- It's as indexable as github discussions
- its also real time chat
- functionally is pretty brilliant
- its been around a long time
- It is also open source
- best mix of realtime communication but with first-class support for topic-based communication

Cons:
- UI isn't exactly stellar

Just citing @JAForbes @achou11

**Reply by kevinfiol** [reply]
If we go the Zulip route, we may want to look at finding a way to archive the existing Gitter chatroom so that its chatlogs are still accessible and potentially searchable from Zulip. I have no experience with Zulip tools, but I did find this documentation about a BETA feature:

https://zulip.com/help/import-from-gitter

**Comment by CreaturesInUnitards** [comment]
> Cons:
> - UI isn't exactly stellar  
  
It's no worse than Gitter! I'm 💯 re: moving to Zulip.

**Comment by kevinfiol** [comment]
Gonna throw [Matrix](https://matrix.org/) in the hat too.

Benefits:
* Open Source
* It's an open protocol, so you can use different clients such as [Element](https://element.io/)
* Free to start new channel on Matrix homeserver
* Gitter was migrated to Matrix not long ago anyway, so the Gitter channel is already accessible via a Matrix client anyway
* Search works better than Gitter's internal search (depending on the client)
* Many clients allow sign-in with Github/Gitlab

Cons:
* Most clients are incomplete or rough around the edges
* Not very newb-friendly
* Need to be signed in to view channel

**Reply by achou11** [reply]
I like Matrix a lot too, but worried that the current state of clients falls into the same trap as other applications like Slack and Discord where conversations just get lost too easily. Could be wrong but I don't think any of the popular clients support subject/thread-based conversations or the like?

**Reply by kevinfiol** [reply]
Agree, I'm still leaning toward Zulip since it combines the niceties of forums + real-time chat. But I felt we should at least consider Matrix, so I added it to this thread.

**Comment by modir** [comment]
I would like to mention [discourse](https://www.discourse.org/) as well. I use it in many other open source software support websites and as a user it is easy to use. They have free hosting for open source project or I would be willing to host it for mithril on one of our servers.

**Reply by dead-claudia** [reply]
Discourse is nice, but would replace GitHub Discussions, not so much Gitter.

**Comment by pdfernhout** [comment]
Hi Kevin,

As another option, here is something I wrote in 2018 to download all the 
Mithril Gitter chat messages and display them in a (simple) Mithril UI. 
https://github.com/pdfernhout/mithril-gitter-archiver-and-viewer

The tool uses the same gitter API as the tool you linked to uses (which 
requires getting a developer token from gitter). I polished that code up 
a bit today and tested it. While it still works, it could always be 
better -- particularly when downloading more message a second time later 
(which is a bit rough, but doable).

That's just the code to download message yourself. I did not want to 
post the downloaded results from gitter or set up a public server with 
them because I am not sure of licensing issues or how people would feel 
about all that.

I don't currently have an opinion about moving away from gitter -- other 
than that gitter has worked OK for me so far.

I wrote the above back in 2018 in part as a proof-of-concept about using 
Mithril to display its own chat room, and also to make it quicker for me 
to search for discussion about particular topics, and also as a step 
towards better communication tools using Mithril. That viewer can also 
display email mbox files somewhat (like from the Mithril mailing list), 
though again in a rough way.

--Paul Fernhout (pdfernhout.net)
"The biggest challenge of the 21st century is the irony of technologies 
of abundance in the hands of those still thinking in terms of scarcity."

**Comment by pygy** [comment]
So, Zulip it is.

We'll hopefully migrate this week, and import the gitter history into Zulip.

Let's plan the migration so as not to forget anything.

Before the migration, we should:
- [x] Figure how to archive Gitter
- [x] Figure how to mass import into Zulip
- [x] Create a gitter reply bot to send people who end up there to the new chat
- [ ] Create a Zulip archive bot that backs up conversations to a github repo, published as gh pages.
- [ ] Figure out if the archive bot can tolerate the volume of the mass migration, and if not, how to make it avoid choking up
- [ ] Update the docs to warn of possible troubled waters around the migration.
  - [ ] the Web site
  - [ ] the Readme
  - [ ] the Wiki
  - [ ] the awesome-mithril Readme

For the migration proper, we'll have to transfer the data and activate the bots.

After the migration, once things have settled, we'll have to once again edit the docs.

Did I forget anything?

**Comment by pygy** [comment]
So, after setting up Python3, I got [this branch of the import script](https://github.com/BenjamenMeyer/archive-gitter/tree/enhancement_config-file) to work. It can list channels and download several of them in one operation, which is nicer than the original.

Config:

```JS
{
    "token": "My Gitter Token",
	"all": false,
	"rooms": [
        "MithrilJS/mithril.js",
        "MithrilJS/mithril-node-render",
        "MithrilJS/Lobby",
        "MithrilJS/admin-talk"
	]
}
```

Its currently downloading.

**Comment by pygy** [comment]
Mass importing to Zulip is just sending the history to them, they create the new org (and can rename the transitional board we created).

I'm wating for a reply on an ETA on their end, I hadn't realized they were the ones who had to do the work (I read too fast I suppose). This adds some uncertainty to the migration calendar.

Off to setting up the bots.

**Comment by pygy** [comment]
As far as archiving goes, this should be easy: https://github.com/zulip/zulip-archive#running-zulip-archive-as-a-github-action

I'll try to use it as a basis for Gandalf as a reply bot on Gitter, using their API

**Comment by pygy** [comment]
So, the Gitter bot should be replying in the `mithril-node-render` channel. It's set on a cron job, but GH doesn't seem to be exactly on time (or I misunderstood the cron timing format).

**Comment by pygy** [comment]
I moved the gitter reply bot to https://github.com/MithrilJS/gitter-reply-bot, uncomment one line and it will start replying.

I've set up https://github.com/pygy/zulip-backup-experiments AFAICT according to their docs, but it doesn't work yet.

**Comment by pygy** [comment]
So...

> Should we switch away from Gitter ?

We did! Things didn't go exactly according to plan, but the Mithril chat now lives on Zulip.
